File: python/intervels/MeetingRooms.py
'''
Given an array of meeting time interval objects consisting of start and end times [[start_1,end_1],[start_2,end_2],...] (start_i < end_i), determine if a person could add all meetings to their schedule without any conflicts.

Note: (0,8),(8,10) is not considered a conflict at 8

Example 1:
Input: intervals = [(0,30),(5,10),(15,20)]
Output: false
Explanation:
(0,30) and (5,10) will conflict
(0,30) and (15,20) will conflict

Example 2:
Input: intervals = [(5,8),(9,15)]
Output: true
'''
from typing import List
from Interval import Interval

class MeetingRooms:
    def canAttendMeetingsNeet(self, intervals: List[Interval]) -> bool:
        intervals.sort(key = lambda x : x.start)
        for i in range(1, len(intervals)):
            i1 = intervals[i -1]
            i2 = intervals[i]
            if i1.end > i2.start:
                return False
        return True
    def canAttendMeetings(self, intervals: List[Interval]) -> bool:
        if not intervals:
            return True
            
        # Sort intervals by start time
        intervals.sort(key=lambda x: x.start)
        for i in range(1, len(intervals)):
            # If current meeting starts before previous meeting ends, there is a conflict
            if intervals[i].start < intervals[i-1].end:
                return False
                
        return True

if __name__ == "__main__":
    solution = MeetingRooms()
    
    # Test Case 1
    intervals1 = [Interval(0, 30), Interval(5, 10), Interval(15, 20)]
    print(f"Input: {intervals1}")
    print(f"Output: {solution.canAttendMeetingsNeet(intervals1)}") # Expected: False
    print("-" * 20)

    # Test Case 2
    intervals2 = [Interval(5, 8), Interval(9, 15)]
    print(f"Input: {intervals2}")
    print(f"Output: {solution.canAttendMeetingsNeet(intervals2)}") # Expected: True
    print("-" * 20)
    
    # Test Case 3 (Touching intervals - not a conflict)
    intervals3 = [Interval(0, 8), Interval(8, 10)]
    print(f"Input: {intervals3}")
    print(f"Output: {solution.canAttendMeetings(intervals3)}") # Expected: True

File: src/main/java/com/naresh/a_dsalgo/Intervals/problems/MeetingRooms.java
package com.naresh.a_dsalgo.Intervals.problems;

import java.util.Arrays;
import java.util.Comparator;

/**
 * Problem: Meeting Rooms (LeetCode 252)
 * Description: Given an array of meeting time intervals `intervals` where `intervals[i] = [starti, endi]`,
 * determine if a person could attend all meetings.
 */
public class MeetingRooms {

    /**
     * Algorithm: Sort the intervals by their start times. Then, iterate through the sorted intervals
     * and check if any current meeting's start time is before the previous meeting's end time.
     * If such an overlap is found, it's impossible to attend all meetings.
     *
     * @param intervals A 2D array of meeting time intervals `[start, end]`.
     * @return True if a person could attend all meetings, false otherwise.
     */
    public boolean canAttendMeetings(int[][] intervals) {
        // Pattern: Interval Sorting | Time: O(N log N), Space: O(1) (or O(N) for sort depending on implementation)
        if (intervals == null || intervals.length <= 1) return true; // No conflicts with 0 or 1 meeting

        // Sort intervals by their start times
        Arrays.sort(intervals, Comparator.comparingInt(a -> a[0]));

        // Check for overlaps
        for (var i = 1; i < intervals.length; i++) {
            // If the current meeting starts before the previous meeting ends, there's an overlap
            if (intervals[i][0] < intervals[i - 1][1]) {
                return false;
            }
        }
        return true;
    }
    // FAANG Tip: Sorting by start time is a common first step for many interval problems. This problem is a simpler variant of Meeting Rooms II.

    public static void main(String[] args) {
        var solution = new MeetingRooms();

        // Test Case 1: Example - No overlap
        var intervals1 = new int[][]{{0, 30}, {5, 10}, {15, 20}};
        // Expected: false (0-30 overlaps with 5-10 and 15-20)
        System.out.println("Intervals: " + Arrays.deepToString(intervals1) + " -> Can attend all: " + solution.canAttendMeetings(intervals1));

        // Test Case 2: Example - No overlap
        var intervals2 = new int[][]{{7, 10}, {2, 4}};
        // Expected: true
        System.out.println("Intervals: " + Arrays.deepToString(intervals2) + " -> Can attend all: " + solution.canAttendMeetings(intervals2));

        // Test Case 3: Edge case - Empty intervals
        var intervals3 = new int[][]{};
        // Expected: true
        System.out.println("Intervals: " + Arrays.deepToString(intervals3) + " -> Can attend all: " + solution.canAttendMeetings(intervals3));

        // Test Case 4: Edge case - Single interval
        var intervals4 = new int[][]{{1, 5}};
        // Expected: true
        System.out.println("Intervals: " + Arrays.deepToString(intervals4) + " -> Can attend all: " + solution.canAttendMeetings(intervals4));

        // Test Case 5: Overlap at boundary
        var intervals5 = new int[][]{{1, 5}, {5, 10}};
        // Expected: true (meeting ends at 5, next starts at 5, no strict overlap)
        System.out.println("Intervals: " + Arrays.deepToString(intervals5) + " -> Can attend all: " + solution.canAttendMeetings(intervals5));

        // Test Case 6: Multiple overlaps
        var intervals6 = new int[][]{{1, 4}, {2, 5}, {3, 6}};
        // Expected: false
        System.out.println("Intervals: " + Arrays.deepToString(intervals6) + " -> Can attend all: " + solution.canAttendMeetings(intervals6));
    }
}
